base/hashing2.jl

```diff
@@ -167,12 +167,35 @@ hash(x::Float16, h::UInt) = hash(float64(x), h)
 
 ## hashing collections ##
 const hashaa_seed = UInt === UInt64 ? 0x7f53e68ceb575e76 : 0xeb575e76
+const hashrle_seed = UInt == UInt64 ? 0x2aab8909bfea414c : 0xbfea414c
 function hash(a::AbstractArray, h::UInt)
     h += hashaa_seed
     h += hash(size(a))
-    for x in a
-        h = hash(x, h)
+
+    state = start(a)
+    done(a, state) && return h
+    x2, state = next(a, state)
+    done(a, state) && return hash(x2, h)
+
+    x1 = x2
+    while !done(a, state)
+        x1 = x2
+        x2, state = next(a, state)
+        if x2 == x1
+            # For repeated elements, use run length encoding
```

Comment:
Why should all nans be equal to each other? Can't we redefine this elsewhere, e.g. in `isequal`? If we distinguish between +0.0 and -0.0, then we're outside of IEEE's numerical equality anyway. We should be able to just use bitwise equality here as well.

Put differently: Different nans exist for a reason (that we may like or not), and there needs to be a way to distinguish between them. `isequal` would be the natural choice for this.

If someone needs to treat all nans as equal, then this particular place should pay the overhead for calling `isnan` and introducing the special case.